File: population.py
from concurrent.futures import ThreadPoolExecutor, as_completed
from NEAT.genome_history import *
from agent import XORsolver
from NEAT.genome import *

import random
import time

class Population:
    def __init__(self, gh, pop_size) -> None:
        self.pop_size = pop_size

        self.population = []
        for _ in range(pop_size):
            self.population.append(XORsolver(gh))

        self.gh = gh
        self.best_fitness = 0.0

    def _create_offspring(self, parents):
        parent1 = parents[random.randint(0, len(parents) // 10)]
        parent2 = parents[random.randint(0, len(parents) // 10)]
        xor = parent1.mate(parent2)
        xor.brian.mutate()
        return xor

    def reset_thread(self):
        parents = self.population

        parents.sort(key=lambda x: x.fitness, reverse=True)

        with ThreadPoolExecutor() as executor:
            
            futures = []
            for off in self.population:
                futures.append(executor.submit(self._create_offspring, parents))
            
            self.population = []
            for future in as_completed(futures):
                self.population.append(future.result())
        self.best_fitness = 0.0


    def reset(self):
        parents = self.population

        parents.sort(key=lambda x: x.fitness, reverse=True)

        self.population = []

        for i in range(self.pop_size):
            parent1 = parents[random.randint(0, len(parents) // 10)]
            parent2 = parents[random.randint(0, len(parents) // 10)]
            xor = parent1.mate(parent2)
            xor.brian.mutate()
            self.population.append(xor)

        self.best_fitness = 0.0


    def update(self, xor_inputs):
        best_xor = self.population[0]
        for xor in self.population:
            start = time.time()
            total_error = 0
            for inputs, expected in xor_inputs:
                out = xor.predict(inputs)
                total_error += (out[0] - expected) ** 2

            xor.fitness = 1 / (total_error + 1e-6)

            if (xor.fitness > self.best_fitness):
                self.best_fitness = xor.fitness
                best_xor = xor

        return best_xor

File: treejs_test/neat/population.ts
import { GenomeHistory } from "./genome_history.ts";
import { Profiler } from "./profiler.ts";
import { XORsolver } from "./xor_test.ts";

export class Population {
  population: XORsolver[] = []

  global_best: XORsolver
  local_best: XORsolver

  generation = 0

  constructor(
    private genome_history: GenomeHistory,
    private population_size: number,
    private agent: typeof XORsolver
  ) {
    for (let i = 0; i < population_size; i++) {
      this.population.push(new agent(genome_history))
    }

    this.local_best = this.population[0]
    this.global_best = this.population[0]
  }

  public reset() {
    const parents = this.population

    parents.sort((a, b) => b.fitness - a.fitness)

    this.population = []

    this.generation++
    for (let i = 0; i < this.population_size; i++) {
      const parent1 = parents[Math.floor(Math.random() * parents.length)]
      const parent2 = parents[Math.floor(Math.random() * parents.length)]

      const child = parent1.mate(parent2)
      child.generation = this.generation
      child.brian.mutate()
      this.population.push(child)
    }

  }

  public update(inputs) {
    this.local_best = this.population[0]

    this.population.forEach((agent) => {
      const fitness = Profiler.time("update", () => agent.calculate_fitness(inputs))

      if (fitness > this.local_best.fitness) this.local_best = agent
      if (fitness > this.global_best.fitness) this.global_best = agent

    })

    return this.local_best
  }
}